pkg/util/hostname.go

```diff
@@ -121,54 +155,192 @@ func saveHostnameData(cacheHostnameKey string, hostname string, provider string)
 	return hostnameData
 }
 
+// ResolveSourcesWithState will merge two HostnameMap sources, providing the expected
+// resolved output.
+func ResolveSourcesWithState(newSources, stateSources HostnameMap) (HostnameMap, bool) {
+	resolvedSources := HostnameMap{}
+	stateChange := false
+
+	if stateSources == nil {
+		return newSources, true
+	}
+
+	for _, stage := range resolutionPipeline {
+		log.Debug("Getting hostname collected by: %s", stage.provider)
+
+		newH, newOk := newSources[stage.provider]
+		stateH, stateOk := stateSources[stage.provider]
+
+		if !newOk && !stateOk {
+			// missing source
+			continue
+		}
+
+		h := newH
+		if stage.reliable {
+			if newOk {
+				if newH != stateH {
+					stateChange = true
```

Comment:
Not entirely sure I follow here. We need to call `configSourceResolver` (points to `ResolveSourcesWithState`) before the `resolutionPipeline` for loop because we need the sources provided by the `configSourceResolver`. That's where we essentially compare the old state to the new state, and provided the resolved state. In line 300, we just check if there were any changes to the resolved state so that we store it. Maybe I misunderstood your question?

Reply:
If I correctly understood, `ResolveSourcesWithState` returns a boolean indicating whether the state changed and must be saved at https://github.com/DataDog/datadog-agent/pull/4798/files#diff-1d9d99d196299ff9a78c3e928058494aR300. 

Instead of using the bool returned by `ResolveSourcesWithState`,  you can compare `sources` and `state` at https://github.com/DataDog/datadog-agent/pull/4798/files#diff-1d9d99d196299ff9a78c3e928058494aR300. You could then remove the bool returned by   `ResolveSourcesWithState`.

Note: This comment was more relevant with the initial implementation of `ResolveSourcesWithState` which was a little more complex.